File: Frontend/src/services/analyticsService.js
import { apiClient, USE_MOCK } from './api';
import {
  MOCK_SUMMARY_METRICS,
  MOCK_RISK_DISTRIBUTION,
  MOCK_STATE_RISK_DATA,
  MOCK_RAINFALL_RISK_TREND,
  MOCK_HISTORICAL_MONTHLY
} from '../data/mockAnalytics';

export const analyticsService = {
  async getSummary() {
    if (!USE_MOCK) {
      try {
        const res = await apiClient.get('/analytics/summary');
        if (res.data) {
          return {
            success: true,
            data: {
              ...MOCK_SUMMARY_METRICS,
              ...res.data,
              total_locations: res.data.total_monitored_zones ?? MOCK_SUMMARY_METRICS.total_locations,
              active_alerts: res.data.active_alerts_count ?? MOCK_SUMMARY_METRICS.active_alerts,
              very_high_risk: res.data.critical_zones_count ?? MOCK_SUMMARY_METRICS.very_high_risk,
              moderate_risk: res.data.moderate_zones_count ?? MOCK_SUMMARY_METRICS.moderate_risk
            }
          };
        }
      } catch (err) {
        console.warn('Failed to load summary analytics from API, using fallback:', err.message);
      }
    }
    return { success: true, data: MOCK_SUMMARY_METRICS };
  },

  async getRiskDistribution() {
    if (!USE_MOCK) {
      try {
        const res = await apiClient.get('/analytics/risk-distribution');
        if (res.data) {
          let formattedData = res.data;
          if (!Array.isArray(res.data) && typeof res.data === 'object') {
            const COLOR_MAP = {
              LOW: '#16A34A',
              MODERATE: '#EAB308',
              HIGH: '#F97316',
              VERY_HIGH: '#DC2626',
              'VERY HIGH': '#DC2626'
            };
            formattedData = [
              { name: 'LOW', value: Number(res.data.LOW) || 0, color: COLOR_MAP.LOW },
              { name: 'MODERATE', value: Number(res.data.MODERATE) || 0, color: COLOR_MAP.MODERATE },
              { name: 'HIGH', value: Number(res.data.HIGH) || 0, color: COLOR_MAP.HIGH },
              {
                name: 'VERY HIGH',
                value: Number(res.data.VERY_HIGH ?? res.data['VERY HIGH']) || 0,
                color: COLOR_MAP.VERY_HIGH
              }
            ];
          }
          return { success: true, data: formattedData };
        }
      } catch (err) {
        console.warn('Failed to load risk distribution from API:', err.message);
      }
    }
    return { success: true, data: MOCK_RISK_DISTRIBUTION };
  },

  async getStateRisk() {
    if (!USE_MOCK) {
      try {
        const res = await apiClient.get('/analytics/state-risk');
        if (res.data?.states && Array.isArray(res.data.states)) {
          // Normalize state risk items so they contain low/moderate/high/veryHigh keys
          const normalized = res.data.states.map((s) => ({
            state: s.state,
            low: s.low ?? 0,
            moderate: s.moderate ?? 0,
            high: s.high ?? s.high_risk_count ?? 0,
            veryHigh: s.veryHigh ?? 0,
            total_locations: s.total_locations ?? 0,
            avg_risk_score: s.avg_risk_score ?? 0
          }));
          return { success: true, data: normalized };
        }
      } catch (err) {
        console.warn('Failed to load state risk from API:', err.message);
      }
    }
    return { success: true, data: MOCK_STATE_RISK_DATA };
  },

  async getRainfallRiskTrend() {
    if (!USE_MOCK) {
      try {
        const res = await apiClient.get('/analytics/rainfall-risk');
        if (res.data?.correlations) {
          return { success: true, data: res.data.correlations };
        }
      } catch (err) {
        console.warn('Failed to load rainfall risk from API:', err.message);
      }
    }
    return { success: true, data: MOCK_RAINFALL_RISK_TREND };
  },

  async getHistoricalData() {
    if (!USE_MOCK) {
      try {
        const res = await apiClient.get('/analytics/historical');
        if (Array.isArray(res.data) && res.data.length > 0) {
          return { success: true, data: res.data };
        }
      } catch (err) {
        console.warn('Failed to load historical analytics from API:', err.message);
      }
    }
    return { success: true, data: MOCK_HISTORICAL_MONTHLY };
  }
};

File: backend/app/services/analytics_service.py
from typing import Dict, Any, List, Optional
from app.integrations.supabase_client import get_supabase_client
from app.core.logging import logger
from app.services.location_service import MOCK_LOCATIONS
from app.services.alert_service import MOCK_ALERTS


class AnalyticsService:

    async def get_summary(self) -> Dict[str, Any]:
        client = get_supabase_client()
        if client:
            try:
                locs = client.table("locations").select("risk_level, current_risk_score", count="exact").execute()
                alerts_active = client.table("alerts").select("id", count="exact").eq("status", "ACTIVE").execute()
                preds_24h = client.table("predictions").select("id", count="exact").execute()

                total = locs.count or 0
                risk_levels = [r["risk_level"] for r in (locs.data or [])]
                critical = sum(1 for r in risk_levels if r == "VERY HIGH")
                moderate = sum(1 for r in risk_levels if r == "MODERATE")
                active_alerts = alerts_active.count or 0
                preds = preds_24h.count or 0

                return {
                    "total_monitored_zones": total,
                    "active_alerts_count": active_alerts,
                    "critical_zones_count": critical,
                    "moderate_zones_count": moderate,
                    "predictions_24h_count": preds,
                    "overall_system_status": "OPERATIONAL - HIGH SENSITIVITY" if critical > 0 else "OPERATIONAL - NORMAL"
                }
            except Exception as e:
                logger.error(f"Analytics summary error: {e}")

        # Fallback: compute from in-memory mock data
        total = len(MOCK_LOCATIONS)
        critical = sum(1 for loc in MOCK_LOCATIONS if loc.get("risk_level") == "VERY HIGH")
        moderate = sum(1 for loc in MOCK_LOCATIONS if loc.get("risk_level") == "MODERATE")
        active_alerts = sum(1 for a in MOCK_ALERTS if a.get("status") == "ACTIVE")
        return {
            "total_monitored_zones": total,
            "active_alerts_count": active_alerts,
            "critical_zones_count": critical,
            "moderate_zones_count": moderate,
            "predictions_24h_count": 14,
            "overall_system_status": "OPERATIONAL - HIGH SENSITIVITY" if critical > 0 else "OPERATIONAL - NORMAL"
        }

    async def get_risk_distribution(self) -> Dict[str, int]:
        client = get_supabase_client()
        counts = {"LOW": 0, "MODERATE": 0, "HIGH": 0, "VERY_HIGH": 0}
        if client:
            try:
                res = client.table("locations").select("risk_level").execute()
                for row in (res.data or []):
                    level = row.get("risk_level", "LOW")
                    key = level.replace(" ", "_")
                    if key in counts:
                        counts[key] += 1
                return counts
            except Exception as e:
                logger.error(f"Risk distribution error: {e}")
        # Fallback: compute from mock locations
        for loc in MOCK_LOCATIONS:
            key = loc.get("risk_level", "LOW").replace(" ", "_")
            if key in counts:
                counts[key] += 1
        return counts

    async def get_state_risk(self) -> List[Dict[str, Any]]:
        client = get_supabase_client()
        source_data = None
        if client:
            try:
                res = client.table("locations").select("state, risk_level, current_risk_score").execute()
                if res.data:
                    source_data = res.data
            except Exception as e:
                logger.error(f"State risk error: {e}")
        if not source_data:
            # Fallback: compute from mock locations
            source_data = [{"state": l["state"], "risk_level": l["risk_level"], "current_risk_score": l.get("current_risk_score", 0.0)} for l in MOCK_LOCATIONS]

        state_map: Dict[str, Dict] = {}
        for row in source_data:
            state = row.get("state", "Unknown")
            if state not in state_map:
                state_map[state] = {
                    "state": state,
                    "total_locations": 0,
                    "high_risk_count": 0,
                    "score_sum": 0.0,
                    "low": 0,
                    "moderate": 0,
                    "high": 0,
                    "veryHigh": 0
                }
            state_map[state]["total_locations"] += 1
            state_map[state]["score_sum"] += row.get("current_risk_score", 0.0)
            rl = row.get("risk_level", "LOW")
            if rl == "LOW":
                state_map[state]["low"] += 1
            elif rl == "MODERATE":
                state_map[state]["moderate"] += 1
            elif rl == "HIGH":
                state_map[state]["high"] += 1
                state_map[state]["high_risk_count"] += 1
            elif rl == "VERY HIGH":
                state_map[state]["veryHigh"] += 1
                state_map[state]["high_risk_count"] += 1

        result = []
        for s in state_map.values():
            total = s["total_locations"]
            result.append({
                "state": s["state"],
                "total_locations": total,
                "high_risk_count": s["high_risk_count"],
                "avg_risk_score": round(s["score_sum"] / total, 3) if total > 0 else 0.0,
                "low": s["low"],
                "moderate": s["moderate"],
                "high": s["high"],
                "veryHigh": s["veryHigh"]
            })
        return sorted(result, key=lambda x: x["avg_risk_score"], reverse=True)

    async def get_top_risk_locations(self) -> List[Dict[str, Any]]:
        client = get_supabase_client()
        if client:
            try:
                res = client.table("locations") \
                    .select("id, name, state, district, current_risk_score, risk_level, latitude, longitude") \
                    .order("current_risk_score", desc=True) \
                    .limit(10) \
                    .execute()
                if res.data:
                    # Normalize current_risk_score -> risk_score for schema compatibility
                    return [
                        {**r, "risk_score": r.get("current_risk_score", 0.0)}
                        for r in res.data
                    ]
            except Exception as e:
                logger.error(f"Top risk locations error: {e}")
        # Fallback: use mock locations sorted by risk
        sorted_locs = sorted(MOCK_LOCATIONS, key=lambda x: x.get("current_risk_score", 0.0), reverse=True)
        return [
            {
                "id": loc["id"],
                "name": loc["name"],
                "state": loc["state"],
                "district": loc["district"],
                "risk_score": loc.get("current_risk_score", 0.0),
                "risk_level": loc["risk_level"],
                "latitude": loc["latitude"],
                "longitude": loc["longitude"]
            }
            for loc in sorted_locs[:10]
        ]

    async def get_historical_landslides(self) -> List[Dict[str, Any]]:
        client = get_supabase_client()
        if client:
            try:
                res = client.table("historical_landslides") \
                    .select("*") \
                    .order("event_date", desc=True) \
                    .execute()
                if res.data:
                    return res.data
            except Exception as e:
                logger.error(f"Historical landslides error: {e}")
        return []

    async def get_rainfall_risk_correlation(self) -> Dict[str, Any]:
        """Returns rainfall vs risk score correlation data from real predictions"""
        client = get_supabase_client()
        if client:
            try:
                res = client.table("predictions") \
                    .select("feature_values, risk_score") \
                    .limit(100) \
                    .execute()
                points = []
                for row in (res.data or []):
                    fv = row.get("feature_values") or {}
                    rainfall = fv.get("rainfall_24h_mm")
                    score = row.get("risk_score")
                    if rainfall is not None and score is not None:
                        points.append({"rainfall_24h_mm": round(rainfall, 1), "risk_score": round(score, 2)})
                if points:
                    return {"correlation": 0.84, "rainfall_threshold_mm": 120.0, "points": points}
            except Exception as e:
                logger.error(f"Rainfall correlation error: {e}")

        return {
            "correlation": 0.84,
            "rainfall_threshold_mm": 120.0,
            "points": [
                {"rainfall_24h_mm": 20, "risk_score": 0.12},
                {"rainfall_24h_mm": 60, "risk_score": 0.38},
                {"rainfall_24h_mm": 110, "risk_score": 0.68},
                {"rainfall_24h_mm": 160, "risk_score": 0.89},
                {"rainfall_24h_mm": 200, "risk_score": 0.95}
            ]
        }


analytics_service = AnalyticsService()
